bindings/go/oci/repository.go

```diff
@@ -380,29 +476,61 @@ func (repo *Repository) getStore(ctx context.Context, component string, version
 	return reference, store, nil
 }
 
-// UploadResource uploads a resource to the repository.
+// UploadResource uploads a [*descriptor.Resource] to the repository.
 func (repo *Repository) UploadResource(ctx context.Context, target runtime.Typed, res *descriptor.Resource, b blob.ReadOnlyBlob) (err error) {
 	done := log.Operation(ctx, "upload resource", log.IdentityLogAttr("resource", res.ToIdentity()))
 	defer done(err)
 
-	var old accessv1.OCIImage
-	if err := repo.scheme.Convert(res.Access, &old); err != nil {
-		return fmt.Errorf("error converting resource old to OCI image: %w", err)
+	desc, access, err := repo.uploadOCIImage(ctx, res.Access, target, b)
+	if err != nil {
+		return fmt.Errorf("failed to upload resource as OCI image: %w", err)
+	}
+
+	res.Size = desc.Size
+	if res.Digest == nil {
+		res.Digest = &descriptor.Digest{}
+	}
+	if err := internaldigest.Apply(res.Digest, desc.Digest); err != nil {
+		return fmt.Errorf("failed to apply digest to resource: %w", err)
+	}
+	res.Access = access
+
+	return nil
+}
+
+// UploadSource uploads a [*descriptor.Source] to the repository.
+func (repo *Repository) UploadSource(ctx context.Context, target runtime.Typed, src *descriptor.Source, b blob.ReadOnlyBlob) (err error) {
```

Comment:
This alters the passed in source. That's not apparent from the function description nor the name of the function. I don't really like side-effects that need to be discovered by usage.

I can tolerate them if they are well documented however. :) That said, I would be even happier if this just returns the altered source. Then I can use it like `src, err = UploadSource` which makes it clearer that the source will be a different source after the call.

That said, if you want to go with just a documentation I can accept that.

Reply:
I agree fundamentally, I changed the method contracts now since the impact is still bearable